src/command/contract/publish.rs

```diff
@@ -0,0 +1,125 @@
+use clap::{ArgGroup, Parser};
+use serde::Serialize;
+
+use crate::options::{GraphRefOpt, ProfileOpt};
+use crate::utils::client::StudioClientConfig;
+use crate::{RoverOutput, RoverResult};
+
+use rover_client::operations::contract::publish::{self, ContractPublishInput};
+use rover_client::RoverClientError;
+use rover_std::Style;
+
+#[derive(Debug, Serialize, Parser)]
+#[clap(
+    group = ArgGroup::new("include_tags_group")
+        .args(&["include_tag", "no_include_tags"]).required(true).multiple(false),
+    group = ArgGroup::new("exclude_tags_group")
+        .args(&["exclude_tag", "no_exclude_tags"]).required(true).multiple(false),
+    group = ArgGroup::new("hide_unreachable_types_group")
+        .args(&["hide_unreachable_types", "no_hide_unreachable_types"]).required(true).multiple(false)
+)]
+pub struct Publish {
+    #[clap(flatten)]
+    graph: GraphRefOpt,
+
+    #[clap(flatten)]
+    profile: ProfileOpt,
+
+    /// The source variant name for this contract variant. Once set, this cannot be changed.
```

Comment:
i tried specifying `--source-variant` as `averys-supercloud@current` instead of just `current` and got an expected but unfortunately confusing error message as a result:

```console
$ APOLLO_REGISTRY_URL="https://api-staging.apollographql.com/graphql" rover contract publish averys-supercloud-staging@contract --source-variant averys-su
percloud-staging@current --hide-unreachable-types --no-include-tags --no-exclude-tags
Publishing configuration to averys-supercloud-staging@contract using credentials from the staging profile.

error[E040]: While publishing the contract configuration and triggering launch, the following error occurred:
The source variant "averys-supercloud-staging@current" does not exist.
```

we could add in a check if there is an `@` included, and if there is, run `GraphRef::parse`. then we just use the extracted variant name if it parses correctly and the graph id matches the contract graph id you're trying to create

Reply:
Unfortunately, we have some existing legacy variants in production that use the `@` symbol in their variant name, so we can't unambiguously tell whether a string that represents a variant name or a graph ref is one or the other 😢 .

We could have two arguments: one named `--source-variant` and one named `--source-variant-name` (the first taking a graph ref, and the second taking a variant name), and put them in an optional `ArgGroup`. Would that be fine here?

Reply:
hmm - could we just call `GraphRef::parse()` on it directly and see if it spits out a variant? there's a regex in there that could help. I think the existing `--source-variant` argument is perfectly fine and we shouldn't add in `--source-variant-name` in addition, I'm just anticipating that this could be a common mistake (typing the full graph ref instead of just the variant) that we could handle a bit better. Specifically, seeing the text `The source variant "averys-supercloud-staging@current" does not exist.` for a variant that _does_ exist is really confusing.

Reply:
Giving an update here -- we're going to change the backend API to accept both graph refs and variant names for `sourceVariant`.